README.md

# Sucrose

A **Bridge Payment Gateway** for Xendit and any other payment gateway.

<p align="center">
    <img alt="GitHub" src="https://img.shields.io/github/license/evnx32/sucrose?style=flat"></a>
    <img src="https://img.shields.io/github/repo-size/evnx32/sucrose" alt="Repo Size" /></a>
    <img alt="Pub Version" src="https://img.shields.io/pub/v/sucrose">
    <img alt="GitHub top language" src="https://img.shields.io/github/languages/top/evnx32/sucrose">
    <img alt="Codacy grade" src="https://img.shields.io/codacy/grade/50dbfa9009ce4cac8f76064077704738">
</p>



## Available Payment Gateway
- [x]   Xendit (WIP)
- [x]   Midtrans (WIP)

## Installation

```bash
flutter pub add sucrose
```
    
## Getting Started

#### Init Xendit
```dart
import 'package:sucrose/sucrose.dart';
    
final sucrose = Sucrose.initXendit(xenditApiKey:"YOUR_API_KEY");
```

#### Init Midtrans
```dart
import 'package:sucrose/sucrose.dart';
    
final sucrose = Sucrose.initMidtrans(
    midtransApiKey: "YOUR_API_KEY",
    environment: MidtransEnvironment.SANDBOX,
    debug: true,
);

```


## Get Started with Xendit


#### Create Invoice
```dart
  final response = await sucrose.xendit.createInvoice(
    request: XenditInvoiceRequest(
      externalId: "ref 13123213",
      amount: 10000,
      description: "your description goes here",
      payerEmail: "hello@mail.com",
      items: [
        XenditInvoiceItem(
          name: "item 1",
          price: 10000,
          quantity: 1,
          category: "meow",
          url: "https://google.com",
        ),
      ],
    ),
  );
```

## Get Started with Midtrans

#### Create Snap Transaction
```dart
    final response = await sucrose.midtrans.createSnapTransaction(
      request: MidtransSnapTransactionRequest(
        transactionDetails: MidtransTransactionDetail(
          orderId: "Inv-${DateTime.now().millisecondsSinceEpoch}",
          grossAmount: 12500,
        ),
        itemDetails: [
          MidtransItemDetail(
            id: "021930129031",
            name: "Steam Wallet IDR 12K",
            price: 12500,
            quantity: 1,
          ),
        ],
        customerDetails: MidtransCustomerDetails(
          email: "test@mail.com",
          firstName: "Test",
          lastName: "Test",
          phone: "08123456789",
          shippingAddress: MidtransShippingAddress(
            address: "Jalan Teknologi Indonesia",
            city: "Jakarta",
            countryCode: "IDN",
            firstName: "Test",
            lastName: "Test",
            phone: "08123456789",
            postalCode: "12345",
          ),
          billingAddress: MidtransBillingAddress(
            address: "Jalan Teknologi Indonesia",
            city: "Jakarta",
            countryCode: "IDN",
            firstName: "Test",
            lastName: "Test",
            phone: "08123456789",
            postalCode: "12345",
          ),
        ),
      ),
    );
```


## Feedback

If you have any feedback, please reach out to us at evnx32@gmail.com


CHANGELOG.md
## 0.0.1


* TODO: Release BETA.


## 0.0.2


* Adding more features.
* Adding error handler.
* Adding some header.
* Moving some import to `xendit_model.dart` for more better to read.
* Refactoring some variables to private.


## 0.0.3


* Adding more payment method.
* Adding refund on the payment.
* Adding some header.
* Adding some endpoints.
* Add debug selection for showing log.
* Adding some documentation on the code.
* Adding `success_redirect_url`, `failure_redirect_url`, and `cancel_redirect_url` on the Channel Properties.
* Adding `desktop_web_checkout_url`, `mobile_web_checkout_url`, `mobile_deeplink_checkout_url`, `qr_checkout_string` on the Actions
* Adding some `enums`.
* Fix minor bugs.


## 0.0.3+1


* Adding Fixed OTC Payment for Indonesia
* Adding OTC Payment for Philippines
* Adding some documentations


## 0.0.4


* Adding Midtrans Payment Gateway.
* Fix minor bug.

## 0.0.4+1


* Fix missing file on `midtrans_model.dart`.


## 0.0.5

* Adding register card to get `saved_token_id` and other stuff.
* Adding cardless payments like `Akulaku` and `Kredivo`.

example/lib/main.dart
import 'package:sucrose/sucrose.dart';
import 'package:flutter/material.dart';
import 'package:logger/logger.dart';

void main() async {
  final sucrose = Sucrose.initMidtrans(
    midtransApiKey: "YOUR API KEY",
    environment: MidtransEnvironment.SANDBOX,
    debug: true,
  );

  Logger logger = Logger(
    printer: PrettyPrinter(),
  );
  try {
    final response = await sucrose.midtrans.createSnapTransaction(
      request: MidtransSnapTransactionRequest(
        transactionDetails: MidtransTransactionDetail(
          orderId: "Inv-${DateTime.now().millisecondsSinceEpoch}",
          grossAmount: 12500,
        ),
        itemDetails: [
          MidtransItemDetail(
            id: "021930129031",
            name: "Steam Wallet IDR 12K",
            price: 12500,
            quantity: 1,
          ),
        ],
        customerDetails: MidtransCustomerDetails(
          email: "test@mail.com",
          firstName: "Test",
          lastName: "Test",
          phone: "08123456789",
          shippingAddress: MidtransShippingAddress(
            address: "Jalan Teknologi Indonesia",
            city: "Jakarta",
            countryCode: "IDN",
            firstName: "Test",
            lastName: "Test",
            phone: "08123456789",
            postalCode: "12345",
          ),
          billingAddress: MidtransBillingAddress(
            address: "Jalan Teknologi Indonesia",
            city: "Jakarta",
            countryCode: "IDN",
            firstName: "Test",
            lastName: "Test",
            phone: "08123456789",
            postalCode: "12345",
          ),
        ),
      ),
    );
    logger.i(response.toJson());
  } catch (e) {
    if (e is MidtransException) {
      logger.e(e.statusCode);
      logger.e(e.statusMessage);
      logger.e(e.validationMessages);
    } else {
      logger.e(e.toString());
    }
  }

  runApp(const MainApp());
}

class MainApp extends StatefulWidget {
  const MainApp({super.key});

  @override
  State<MainApp> createState() => _MainAppState();
}

class _MainAppState extends State<MainApp> {
  // final controller = WebViewController()
  //   ..setJavaScriptMode(JavaScriptMode.unrestricted)
  //   ..setBackgroundColor(const Color(0x00000000))
  //   ..setNavigationDelegate(
  //     NavigationDelegate(
  //       onProgress: (int progress) {
  //         // Update loading bar.
  //       },
  //       onPageStarted: (String url) {},
  //       onPageFinished: (String url) {},
  //       onWebResourceError: (WebResourceError error) {},
  //       onNavigationRequest: (NavigationRequest request) {
  //         if (request.url.startsWith('https://www.youtube.com/')) {
  //           return NavigationDecision.prevent;
  //         }
  //         return NavigationDecision.navigate;
  //       },
  //     ),
  //   )
  //   ..loadRequest(Uri.parse(
  //       'https://app.sandbox.midtrans.com/snap/v3/redirection/31ccb075-2c7e-4af8-a0b1-1f8ac1dea6a6'));
  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      home: Scaffold(body: Center(child: Text("Hello"))),
    );
  }
}
